### PR TITLE
fix issue 262: close peers that have no addresses in the host.peerstore due to expired ttl.

### PR DESCRIPTION
**Pull request by ping-ke**
issue 262: https://github.com/ethstorage/es-node/issues/262

**Root Cause**
When creating a newStream to send a request to a peer, if the connection does not exist, it will call the peer again. But if the address of that peer does not exist, this issue will be thrown.
The peer address does not exist because of time-to-live timeout, which is 24 hours correctly. time-to-live timeout means the peer not be added back to peerstore by discovery.

pstore.AddAddrs(info.ID, info.Addrs, discoveredAddrTTL) // discoveredAddrTTL = time.Hour * 24
 

**How to fix it**
Change the log level to debug for the error caused by newStream "no addresses" and close the peer.

### CODE REVIEW

**Comment by qzhodl** [comment]
> the peer will remove or recover after some time

Why is the peer removed or recovered after some time? Could you please provide more explanation?

**Comment by qzhodl** [comment]
I am still a little concerned because we haven’t identified the root cause of addresses disappearing in the peer store. If we change this error to the debug level, we will essentially lose visibility of the issue.

**Comment by qzhodl** [comment]
> Change the log level to debug for the error caused by newStream, this should be a temporary issue between the local node and the remote peer, I have seen this issue twice (one reported by a customer and one by me - the network condition at that time was very bad.)
> 
> This issue will be resolved some times later when
> 
> 1. When the peer is removed
> 2. Addr is added back to Peerstore in the discovery.go

According to the discussion in today’s meeting, I think we may need to set up a test environment to determine if the root cause is an outdated TTL.

**Comment by ping-ke** [comment]
Added log to removeExpired in addr_book.go, and will see it clear the addresses after expired. and connect to this peer will throw 
"no addresses" issue.

**Comment by qzhodl** [comment]
> Added log to removeExpired in addr_book.go, and will see it clear the addresses after expired. and connect to this peer will throw "no addresses" issue.
> 
> 

Then should we also need to remove the peer in syncClient when the peerstore removes the addr?

File: ethstorage/p2p/protocol/syncclient.go [hunk 1]
```diff
@@ -731,7 +732,10 @@ func (s *SyncClient) assignBlobRangeTasks() {
 
 				if err != nil {
```
Comment (syntrust) [inline ethstorage/p2p/protocol/syncclient.go:733]: Should the error be processed immediately after `err` is got?

File: ethstorage/p2p/protocol/syncclient.go [hunk 2]
```diff
@@ -1260,6 +1267,40 @@ func (s *SyncClient) ReportPeerSummary() {
 	}
 }
 
+func (s *SyncClient) PurgeBadPeers(closePeerFunc func(peer.ID) error) {
```
Comment (syntrust) [inline ethstorage/p2p/protocol/syncclient.go:1270]: Why the peer cannot be closed by `RemovePeer()`? And does this operation need a metrics update?

Comment (ping-ke) [reply ethstorage/p2p/protocol/syncclient.go:1270]: RemovePeer is the function that removes a peer from sync client, and closePeerFunc will close the peer connection, and the RemovePeer() will be called after the connection closed.

File: ethstorage/p2p/protocol/syncclient.go [hunk 3]
```diff
@@ -1260,6 +1267,40 @@ func (s *SyncClient) ReportPeerSummary() {
...
+			}
+		case <-s.resCtx.Done():
+			log.Info("P2P PurgeBadPeers stop")
+			return
+		}
+	}
+}
+
+// TODO we can add more rules for bad peers
+func (s *SyncClient) getBadPeers() []peer.ID {
```
Comment (qzhodl) [inline ethstorage/p2p/protocol/syncclient.go:1292]: The PR description needs to be updated if the changes are more than just lowering the log level.

File: ethstorage/p2p/protocol/syncclient.go [hunk 4]
```diff
@@ -731,7 +732,10 @@ func (s *SyncClient) assignBlobRangeTasks() {
 
 				if err != nil {
 					if e, ok := err.(*yamux.Error); ok && e.Timeout() {
-						log.Debug("Failed to request blobs", "peer", pr.id.String(), "err", err)
+						log.Debug("Request blobs timeout", "peer", pr.id.String(), "err", err)
+					} else if returnCode == streamError && strings.Contains(err.Error(), "no addresses") {
+						pr.isBad = true
```
Comment (qzhodl) [inline ethstorage/p2p/protocol/syncclient.go:737]: Does Prysm have a similar issue to ours, and how do they handle it?

Comment (ping-ke) [reply ethstorage/p2p/protocol/syncclient.go]: Prysm does not add addresses to peerStore, but directly connect to peer using AddrInfo.
a343d35f97d364847231b9798a7e4cb
6710635e392f82458cdc41cb35fe7bd

File: ethstorage/p2p/node.go [hunk 5]
```diff
@@ -182,10 +185,36 @@ func (n *NodeP2P) init(resourcesCtx context.Context, rollupCfg *rollup.EsConfig,
 		if m != nil {
 			go m.RecordBandwidth(resourcesCtx, bwc)
 		}
+
+		go n.PurgeBadPeers()
 	}
 	return nil
 }
 
+func (n *NodeP2P) PurgeBadPeers() {
```
Comment (qzhodl) [inline ethstorage/p2p/node.go:194]: Need a comment to explain why the addresses might disappear

**Review by qzhodl — APPROVED** [review]

**Review by syntrust — APPROVED** [review]

**Pull request merged**